### PR TITLE
Process Emergency Stop

### PR DESCRIPTION
**Pull request by sithlord48**
Emergency Stop is now no the testGUI and when pressed it will either send M112 directly or if the printer is in a busy state it will set the printers state to "STOP" and atcore will do the rest.

### CODE REVIEW

File: testclient/mainwindow.cpp [hunk 1]
```diff
@@ -263,6 +264,15 @@ void MainWindow::printPBClicked()
     }
 }
 
+void MainWindow::stopPBClicked()
+{
+    if (core->state() == BUSY) {
+        core->setState(STOP);
+        ui->printPB->setText(tr("Print File"));
+    } else {
+        core->serial()->pushCommand(QString("M112").toLocal8Bit());
```
Comment (patrickelectric) [inline testclient/mainwindow.cpp]: Here, the command need to be sent to atcore.

File: testclient/mainwindow.cpp [hunk 2]
```diff
@@ -263,6 +264,15 @@ void MainWindow::printPBClicked()
     }
 }
 
+void MainWindow::stopPBClicked()
+{
+    if (core->state() == BUSY) {
```
Comment (patrickelectric) [inline testclient/mainwindow.cpp]: It's better to send only the stop command to the printer. Don't have problem if you send more than one.

File: src/atcore.cpp [hunk 3]
```diff
@@ -172,6 +172,10 @@ void AtCore::print(const QString &fileName)
 
         else if (printerState == STOP) {
             qDebug() << tr("Stop State");
+            //slip stop command into the gcodestream. Let the firmware do the rest.
+            QString stopString("M112");
```
Comment (patrickelectric) [inline src/atcore.cpp]: use gcode namespace

File: testclient/mainwindow.cpp (hunk 1 above)
Comment (patrickelectric) [reply testclient/mainwindow.cpp]: use gcode namespace

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: src/atcore.cpp (hunk 3 above)
Comment (sithlord48) [reply src/atcore.cpp]: :heavy_check_mark:

File: testclient/mainwindow.cpp (hunk 1 above)
Comment (sithlord48) [reply testclient/mainwindow.cpp]: :heavy_check_mark:

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Use setState and state functions

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Change from if else to switch case

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Change from d->serial to serial()

**Review by patrickelectric — APPROVED** [review]

**Pull request closed without merging**